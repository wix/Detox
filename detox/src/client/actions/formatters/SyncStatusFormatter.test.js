const format = require('./SyncStatusFormatter');

describe('Sync Status Formatter', () => {
  describe('assertions', () => {
    it('should throw error when `app_status` is missing', async () => {
      let invalidStatus = {
        busy_resource: []
      };

      await expect(() => {
        format(invalidStatus);
      }).toThrowErrorMatchingSnapshot();
    });

    it('should throw error when `app_status` is invalid', async () => {
      let invalidStatus = {
        app_status: 'foo'
      };

      await expect(() => {
        format(invalidStatus);
      }).toThrowErrorMatchingSnapshot();
    });

    it('should throw error when `app_status` is `busy` but `busy_resources` is missing', async () => {
      let invalidStatus = {
        app_status: 'busy'
      };

      await expect(() => {
        format(invalidStatus);
      }).toThrowErrorMatchingSnapshot();
    });

    it('should throw error when `app_status` is `busy` but `busy_resources` is empty', async () => {
      let invalidStatus = {
        app_status: 'busy',
        busy_resources: []
      };

      await expect(() => {
        format(invalidStatus);
      }).toThrowErrorMatchingSnapshot();
    });

    it('should throw error when resource `name` is missing', async () => {
      let invalidStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            description: {
              foo: 'bar'
            }
          }
        ]
      };

      await expect(() => {
        format(invalidStatus);
      }).toThrowErrorMatchingSnapshot();
    });

    it('should throw error when a busy resource is invalid', async () => {
      let invalidStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'foo'
          }
        ]
      };

      await expect(() => {
        format(invalidStatus);
      }).toThrowErrorMatchingSnapshot();
    });
  });

  it('should format idle status correctly', async () => {
    let idleStatus = {
      app_status: 'idle'
    };

    await expect(format(idleStatus)).toMatchSnapshot();
  });

  describe('busy status', () => {
    it('should format "delayed_perform_selector" correctly', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'delayed_perform_selector',
            description: {
              pending_selectors: 123
            }
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "dispatch_queue" correctly', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'dispatch_queue',
            description: {
              queue: 'foo',
              works_count: 123
            }
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "run_loop" correctly', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'run_loop',
            description: {
              name: 'foo'
            }
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "one_time_events" correctly when there is no object', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'one_time_events',
            description: {
              event: 'foo'
            }
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "one_time_events" correctly', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'one_time_events',
            description: {
              event: 'foo',
              object: 'bar'
            }
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "timers" correctly when there is no description', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'timers'
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "timers" correctly when there are timers in description', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'timers',
            description: {
              timers: [
                {
                  fire_date: 'foo',
                  time_until_fire: 0.468978,
                  repeat_interval: 1,
                  is_recurring: true
                },
                {
                  fire_date: 'bar',
                  time_until_fire: 0.98798,
                  repeat_interval: 0,
                  is_recurring: false
                }
              ]
            }
          }

        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "ui" correctly #1', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'ui',
            description: {
              layer_animation_pending_count: 3,
              layer_needs_display_count: 5,
              layer_needs_layout_count: 1,
              layer_pending_animation_count: 12
            }
          }

        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "ui" correctly #2', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'ui',
            description: {
              view_animation_pending_count: 3,
              view_controller_will_appear_count: 5,
              view_controller_will_disappear_count: 1,
              view_needs_display_count: 12,
              view_needs_layout_count: 15
            }
          }

        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "ui" correctly #3', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'ui',
            description: {
              reason: 'foo'
            }
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "network" correctly', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'network',
            description: {
              urls: [
                'foo://bar.baz',
                'qux://quux.quuz'
              ]
            }
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "js_timers" corrrectly', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'js_timers',
            description: {
              timers: [
                {
                  timer_id: 4,
                  duration: 1,
                  is_recurring: false
                },
                {
                  timer_id: 12,
                  duration: 2,
                  is_recurring: true
                }
              ]
            }
          }

        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "looper" correctly #1', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'looper',
            description: {
              thread: 'Foo Looper',
              execution_type: 'baz execution'
            }
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "looper" correctly', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'looper',
            description: {
              thread: 'Foo Looper'
            }
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "io" correctly', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'io'
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "bridge" correctly', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'bridge'
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "unknown" correctly', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'unknown',
            description: {
              identifier: 'foo.bar#baz'
            }
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });

    it('should format "bg" correctly', async () => {
      let busyStatus = {
        app_status: 'busy',
        busy_resources: [
          {
            name: 'bg',
            description: {
              reason: 'asynctasks'
            }
          }
        ]
      };

      await expect(format(busyStatus)).toMatchSnapshot();
    });
  });
});
