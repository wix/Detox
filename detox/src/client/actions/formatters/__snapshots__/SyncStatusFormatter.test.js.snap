// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sync Status Formatter assertions should throw error when \`app_status\` is \`busy\` but \`busy_resources\` is empty 1`] = `
"Given sync status is not compatible with the status schema (\`SyncStatusSchema.js\`), given status: {\\"app_status\\":\\"busy\\",\\"busy_resources\\":[]}.
With reasons:
• must NOT have additional properties in path \\"#/oneOf/0/additionalProperties\\" with params: {\\"additionalProperty\\":\\"busy_resources\\"}
• must NOT have fewer than 1 items in path \\"#/oneOf/1/properties/busy_resources/minItems\\" with params: {\\"limit\\":1}
• must match exactly one schema in oneOf in path \\"#/oneOf\\" with params: {\\"passingSchemas\\":null}

Please report this issue on our GitHub tracker:
https://github.com/wix/Detox/issues"
`;

exports[`Sync Status Formatter assertions should throw error when \`app_status\` is \`busy\` but \`busy_resources\` is missing 1`] = `
"Given sync status is not compatible with the status schema (\`SyncStatusSchema.js\`), given status: {\\"app_status\\":\\"busy\\"}.
With reasons:
• must be equal to constant in path \\"#/oneOf/0/properties/app_status/const\\" with params: {\\"allowedValue\\":\\"idle\\"}
• must have required property 'busy_resources' in path \\"#/oneOf/1/required\\" with params: {\\"missingProperty\\":\\"busy_resources\\"}
• must match exactly one schema in oneOf in path \\"#/oneOf\\" with params: {\\"passingSchemas\\":null}

Please report this issue on our GitHub tracker:
https://github.com/wix/Detox/issues"
`;

exports[`Sync Status Formatter assertions should throw error when \`app_status\` is invalid 1`] = `
"Given sync status is not compatible with the status schema (\`SyncStatusSchema.js\`), given status: {\\"app_status\\":\\"foo\\"}.
With reasons:
• must be equal to constant in path \\"#/oneOf/0/properties/app_status/const\\" with params: {\\"allowedValue\\":\\"idle\\"}
• must have required property 'busy_resources' in path \\"#/oneOf/1/required\\" with params: {\\"missingProperty\\":\\"busy_resources\\"}
• must match exactly one schema in oneOf in path \\"#/oneOf\\" with params: {\\"passingSchemas\\":null}

Please report this issue on our GitHub tracker:
https://github.com/wix/Detox/issues"
`;

exports[`Sync Status Formatter assertions should throw error when \`app_status\` is missing 1`] = `
"Given sync status is not compatible with the status schema (\`SyncStatusSchema.js\`), given status: {\\"busy_resource\\":[]}.
With reasons:
• must have required property 'app_status' in path \\"#/oneOf/0/required\\" with params: {\\"missingProperty\\":\\"app_status\\"}
• must have required property 'app_status' in path \\"#/oneOf/1/required\\" with params: {\\"missingProperty\\":\\"app_status\\"}
• must match exactly one schema in oneOf in path \\"#/oneOf\\" with params: {\\"passingSchemas\\":null}

Please report this issue on our GitHub tracker:
https://github.com/wix/Detox/issues"
`;

exports[`Sync Status Formatter assertions should throw error when a busy resource is invalid 1`] = `
"Given sync status is not compatible with the status schema (\`SyncStatusSchema.js\`), given status: {\\"app_status\\":\\"busy\\",\\"busy_resources\\":[{\\"name\\":\\"foo\\"}]}.
With reasons:
• must NOT have additional properties in path \\"#/oneOf/0/additionalProperties\\" with params: {\\"additionalProperty\\":\\"busy_resources\\"}
• must have required property 'description' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/0/required\\" with params: {\\"missingProperty\\":\\"description\\"}
• must have required property 'description' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/1/required\\" with params: {\\"missingProperty\\":\\"description\\"}
• must have required property 'description' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/2/required\\" with params: {\\"missingProperty\\":\\"description\\"}
• must have required property 'description' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/3/required\\" with params: {\\"missingProperty\\":\\"description\\"}
• must be equal to constant in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/4/properties/name/const\\" with params: {\\"allowedValue\\":\\"timers\\"}
• must have required property 'description' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/5/required\\" with params: {\\"missingProperty\\":\\"description\\"}
• must have required property 'description' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/6/required\\" with params: {\\"missingProperty\\":\\"description\\"}
• must have required property 'description' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/7/required\\" with params: {\\"missingProperty\\":\\"description\\"}
• must have required property 'description' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/8/required\\" with params: {\\"missingProperty\\":\\"description\\"}
• must have required property 'description' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/9/required\\" with params: {\\"missingProperty\\":\\"description\\"}
• must be equal to constant in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/10/properties/name/const\\" with params: {\\"allowedValue\\":\\"io\\"}
• must be equal to constant in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/11/properties/name/const\\" with params: {\\"allowedValue\\":\\"bridge\\"}
• must have required property 'description' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/12/required\\" with params: {\\"missingProperty\\":\\"description\\"}
• must match a schema in anyOf in path \\"#/oneOf/1/properties/busy_resources/items/anyOf\\" with params: {}
• must match exactly one schema in oneOf in path \\"#/oneOf\\" with params: {\\"passingSchemas\\":null}

Please report this issue on our GitHub tracker:
https://github.com/wix/Detox/issues"
`;

exports[`Sync Status Formatter assertions should throw error when resource \`name\` is missing 1`] = `
"Given sync status is not compatible with the status schema (\`SyncStatusSchema.js\`), given status: {\\"app_status\\":\\"busy\\",\\"busy_resources\\":[{\\"description\\":{\\"foo\\":\\"bar\\"}}]}.
With reasons:
• must NOT have additional properties in path \\"#/oneOf/0/additionalProperties\\" with params: {\\"additionalProperty\\":\\"busy_resources\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/0/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/1/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/2/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/3/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/4/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/5/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/6/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/7/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/8/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/9/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/10/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/11/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must have required property 'name' in path \\"#/oneOf/1/properties/busy_resources/items/anyOf/12/required\\" with params: {\\"missingProperty\\":\\"name\\"}
• must match a schema in anyOf in path \\"#/oneOf/1/properties/busy_resources/items/anyOf\\" with params: {}
• must match exactly one schema in oneOf in path \\"#/oneOf\\" with params: {\\"passingSchemas\\":null}

Please report this issue on our GitHub tracker:
https://github.com/wix/Detox/issues"
`;

exports[`Sync Status Formatter busy status should format "bg" correctly 1`] = `
"The app is busy with the following tasks:
• Background work taking place in asynctasks."
`;

exports[`Sync Status Formatter busy status should format "bridge" correctly 1`] = `
"The app is busy with the following tasks:
• Activity on the React-Native bridge."
`;

exports[`Sync Status Formatter busy status should format "delayed_perform_selector" correctly 1`] = `
"The app is busy with the following tasks:
• There are 123 pending delayed selectors to be performed."
`;

exports[`Sync Status Formatter busy status should format "dispatch_queue" correctly 1`] = `
"The app is busy with the following tasks:
• There are 123 work items pending on the dispatch queue: \\"foo\\"."
`;

exports[`Sync Status Formatter busy status should format "io" correctly 1`] = `
"The app is busy with the following tasks:
• Disk I/O activity."
`;

exports[`Sync Status Formatter busy status should format "js_timers" corrrectly 1`] = `
"The app is busy with the following tasks:
• 2 enqueued JavaScript timers:
  - Timer #1:
    + JS timer ID: 4.
    + Duration: 1.
    + Is recurring: NO.
  - Timer #2:
    + JS timer ID: 12.
    + Duration: 2.
    + Is recurring: YES."
`;

exports[`Sync Status Formatter busy status should format "looper" correctly #1 1`] = `
"The app is busy with the following tasks:
• Foo Looper is executing (baz execution)."
`;

exports[`Sync Status Formatter busy status should format "looper" correctly 1`] = `
"The app is busy with the following tasks:
• Foo Looper is executing."
`;

exports[`Sync Status Formatter busy status should format "network" correctly 1`] = `
"The app is busy with the following tasks:
• 2 network requests with URLs:
  - URL #1: foo://bar.baz
  - URL #2: qux://quux.quuz"
`;

exports[`Sync Status Formatter busy status should format "one_time_events" correctly 1`] = `
"The app is busy with the following tasks:
• The event \\"foo\\" is taking place with object: \\"bar\\"."
`;

exports[`Sync Status Formatter busy status should format "one_time_events" correctly when there is no object 1`] = `
"The app is busy with the following tasks:
• The event \\"foo\\" is taking place."
`;

exports[`Sync Status Formatter busy status should format "run_loop" correctly 1`] = `
"The app is busy with the following tasks:
• Run loop \\"foo\\" is awake."
`;

exports[`Sync Status Formatter busy status should format "timers" correctly when there are timers in description 1`] = `
"The app is busy with the following tasks:
• 2 enqueued native timers:
  - Timer #1:
    + Fire date: foo.
    + Time until fire: 0.469.
    + Repeat interval: 1.
    + Is recurring: YES.
  - Timer #2:
    + Fire date: bar.
    + Time until fire: 0.988.
    + Repeat interval: 0.
    + Is recurring: NO."
`;

exports[`Sync Status Formatter busy status should format "timers" correctly when there is no description 1`] = `
"The app is busy with the following tasks:
• There are enqueued timers."
`;

exports[`Sync Status Formatter busy status should format "ui" correctly #1 1`] = `
"The app is busy with the following tasks:
• UI elements are busy:
  - Layer animations pending: 3.
  - Layers needs display: 5.
  - Layers needs layout: 1.
  - Layers pending animations: 12."
`;

exports[`Sync Status Formatter busy status should format "ui" correctly #2 1`] = `
"The app is busy with the following tasks:
• UI elements are busy:
  - View animations pending: 3.
  - View controllers will appear: 5.
  - View controllers will disappear: 1.
  - View needs display: 12.
  - View needs layout: 15."
`;

exports[`Sync Status Formatter busy status should format "ui" correctly #3 1`] = `
"The app is busy with the following tasks:
• UI elements are busy:
  - Reason: foo."
`;

exports[`Sync Status Formatter busy status should format "unknown" correctly 1`] = `
"The app is busy with the following tasks:
• Resource \\"foo.bar#baz\\" is busy."
`;

exports[`Sync Status Formatter should format idle status correctly 1`] = `"The app seems to be idle"`;
